src/accelerate/utils/dataclasses.py

```diff
@@ -1284,6 +1287,22 @@ def set_moe_leaf_modules(self, model):
                     transformer_moe_cls.append(transformer_cls)
             set_z3_leaf_modules(model, transformer_moe_cls)  # z3_leaf
 
+    def enable(self):
```

Comment:
We've toyed with such APIs, and they don't make sense with how DS works right now. 

I also don't think we should fully tie it to the Accelerator, as the accelerator is a "generic" API for things, this is hyper-specific for DeepSpeed. 

Re; users may call `.enabled=True`, I'll put some blockers in place to avoid users doing the wrong thing. 

Re; when to set it, yeah we probably should move this to the `AcceleratorState` itself.

Re; dict: on paper, that's fine. But it's a very new pattern which breaks backward compatibility, and the overall API with how things are **currently** makes it very clunky, because we cannot prepare multiple models/engines at once with a singular Accelerator.

Once we can do that (aka no need for two `Accelerator`'s to do two models), I am 100% behind `dict` support.

Reply:
> Re; dict: on paper, that's fine. But it's a very new pattern which breaks backward compatibility, and the overall API with how things are currently makes it very clunky, because we cannot prepare multiple models/engines at once with a singular Accelerator. 

It'd work if you make it dual: list and dict. The init 2+ plugins is new anyway, so you can have it init'ed as a dict - the bonus is that it'd be much easier to retrieve the right plugin from the Accelerator and not requiring the user to have their own object they need to pass through to many calls.

So you can return a list of plugins, or a dict, or a specific plugin by its key.

Reply:
What I'm stuck on is what is the key if the user doesn't pass in a dict? Do we do them ordered as the key? (so key `0` gets the first plugin, key `1` gets the second, ...)

Reply:
since the original proposal of yours didn't enforce or prescribe any particular order, then the sorting doesn't matter as long as it's consistent. So just using the standard `d.values()` should sort it consistently.

As I suggested elsewhere, I think `deepspeed_plugins` is unlikely to be used since what would be the point if you don't know what they are - how would you know plugin is which?

But `deepspeed_plugin(key="teacher")` vs `deepspeed_plugin(key="student")` is very useful and most likely will be the way users will retrieve the desired plugin in their code.

Based on that probably just using the dict interface is sufficient and there is no need for a list at all.

This is a new API, so I'm not sure why you said it'd be a BC issue. If `deepspeed_plugin` continues returning the active plugin then everything should just work as before with a single plugin.

before w/ a single plugin:
```
accelerate.deepspeed_plugin().foo()
```
after w/ a single plugin:
```
accelerate.deepspeed_plugin().foo() # returns active plugin
```
after w/ multiple plugins:
```
accelerate.deepspeed_plugin(key="teacher").foo() # returns named plugin
accelerate.deepspeed_plugin(key="student").foo() # returns named plugin
```
and as I suggested elsewhere a magical accessor will make it even nicer:
```
accelerate.deepspeed_plugin.teacher.foo()
accelerate.deepspeed_plugin.student.foo()
```

so the only problem with the last one is that I'm using the same `deepspeed_plugin` as an attribute and as a method - so I don't think this would work and would require a different accessor name.

Reply:
This would wind up being quite a bit breaking, as all HF codebases that use accelerate for DS do `accelerate.deepspeed_plugin` rather than `accelerate.deepspeed_plugin()`.  (and examples, etc)

If we go functional, I'd rather do `accelerator.deepspeed_plugins(key="teacher")`.

*Or* this waits until after 1.0 for this particular side of the API. Are you willing to wait a few weeks before introducing it? :)

Reply:
`accelerator.deepspeed_plugins(key="teacher")` always returning a singular but named plural is not ideal.

ok, what about any plugin object knowing how to get other objects? e.g. 
```
accelerate.deepspeed_plugin.teacher.foo()
```
so it'd `getattr` and find `teacher` in `deepspeed_plugins`? Too magical?

> Or this waits until after 1.0 for this particular side of the API. Are you willing to wait a few weeks before introducing it? 

Personally I'm totally not in a rush, I'm ramping up on inference at the moment.

Reply:
so:
```
# return current active plugin
accelerate.deepspeed_plugin

# return a specific named plugin
accelerate.deepspeed_plugin.teacher
accelerate.deepspeed_plugin.student

Reply:
Feels a bit too magical. I don't expect (without being explicit) the `deepspeed_plugin` to interact with the rest. If it were `deepspeed_plugins` then, and only then, I would. 

Or: `deepspeed_plugin_container` if we wanted to get very explicit.

Reply:
Even with being explicit, I'm still not sure such magic should exist in Accelerate, which is aimed to be non magical as much as possible (even if convenient).

Requiring users maintain their `DeepSpeedPlugins` IMO isn't a terrible requirement when working on multi-plugin based tasks, + a k/v system to access if needed.

The `Accelerator` shouldn't have something like `accelerator.get_deepspeed_plugin(name)` however I could be okay with:

```python
accelerator.state.get_deepspeed_plugin(key)
```

Reply:
> Requiring users maintain their DeepSpeedPlugins IMO isn't a terrible requirement when working on multi-plugin based tasks, + a k/v system to access if needed.

I disagree here, because it then requires bouncing yet another function argument between potentially many calls. Being able to get it from state which is a global is a way easier.

> `accelerator.state.get_deepspeed_plugin(key)`

works for me.

Reply:
as I suggested earlier I recommend asking those users who asked for this feature to kick the tires once you are happy with your work and see how well it works for them.

The biggest challenge here is to think of the various use cases and ensure they are all catered to.

Reply:
Okay pushed new changes, I think this fixes most of the code smell for you @BenjaminBossan. We still need to check `self`, but users can't do `disable()` now, they *must* pick an active plugin to use at all times, and just swap the active one out.

Reply:
@stas00 lmk if you have any other thoughts post seeing the new push

Reply:
I like the new changes, but since I've never had to deal with this use case, please don't put too much weight on my perspective here :)

One thing I would have implemented differently is that I would have had something like a `_active_deepspeed_plugin_key` or so on the `state` instead of an `enabled` attribute on the plugin. The `_active_deepspeed_plugin_key` would always be one of the keys from the dict (or `None` if there is no DS plugin). This would be in line with the idea of making "illegal states unrepresentable", i.e. it would be impossible to have multiple enabled DS plugins or that none are enabled.

Maybe there is a good reason not to do it like this that I'm missing (say, in the future we want to allow multiple active DS plugins), in that case please ignore my comment.